<a href="https://colab.research.google.com/github/pia-francesca/ema/blob/main/ema_tool_application_example.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Application example | **Embedding of protein families across protein language models**

This notebook demonstrates how to use the ema-tool library to analyse the embedding of protein families across protein language models. We will use the [ESM Model](https://github.com/facebookresearch/esm) model and a set of ion channel wildtype proteins as an example.

## Loading dependencies

In [1]:
 ! git clone https://github.com/pia-francesca/ema.git

Cloning into 'ema'...
remote: Enumerating objects: 288, done.
remote: Counting objects: 100% (68/68), done.
remote: Compressing objects: 100% (43/43), done.
remote: Total 288 (delta 29), reused 39 (delta 17), pack-reused 220
Receiving objects: 100% (288/288), 70.30 MiB | 6.67 MiB/s, done.
Resolving deltas: 100% (136/136), done.


In [2]:
! pip install umap

  Preparing metadata (setup.py) ... done
  Created wheel for umap: filename=umap-0.1.1-py3-none-any.whl size=3543 sha256=61d2eb6e3586f0b0bacc5be3fcf2db4ad3583cda2c59f00ec9b5ab65eaf247f9
  Stored in directory: /root/.cache/pip/wheels/15/f1/28/53dcf7a309118ed35d810a5f9cb995217800f3f269ab5771cb
Successfully built umap


In [3]:
import numpy as np
import pandas as pd

from ema.ema.ema import EmbeddingHandler

## Loading data

`ema-tool` requires a table of **metadata** about the proteins to be embedded. The first column of the table should contain the sample identifiers. These need to be unique. Any further columns can contain metadata about the samples. In this example, we will use a table of ion channel proteins. Metadata can be either categorical or numerical.

Further the **embedding spaces** of the proteins need to be provided in the format of NumPy arrays. In this example, we will use the ESM-1b model to embed the proteins. Each row in the NumPy array should correspond to the embedding of the protein with the same row index in the metadata table. The dimensions of the embedding space can differ between models.

In [8]:
DATA_DIR = 'ema/examples/HCN1-variants/'
FP_METADATA = DATA_DIR + 'metadata.csv'
FP_EMB_ESM1b = DATA_DIR + 'esm1b_t33_650M_UR50S-embeddings.npy'
FP_EMB_ESM2 = DATA_DIR + 'esm2_t33_650M_UR50D-embeddings.npy'

In [9]:
# load metadata and embeddings

metadata = pd.read_csv(FP_METADATA)
emb_esm1b = np.load(FP_EMB_ESM1b)
emb_esm2 = np.load(FP_EMB_ESM2)

print(emb_esm1b.shape, emb_esm2.shape)
metadata.head()

(881, 1280) (881, 1280)


variant_id  disorder_propensity ('uniprot', 'Topological domain')  \
0  HCN1:272:S:P                0.010                             False   
1  HCN1:279:H:Y                0.015                             False   
2  HCN1:391:G:D                0.005                             False   
3  HCN1:153:M:I                0.009                             False   
4  HCN1:305:M:L                0.006                             False   

  ('uniprot', 'Transmembrane') ('uniprot', 'Intramembrane')  \
0                         True                        False   
1                         True                        False   
2                         True                        False   
3                         True                        False   
4                         True                        False   

  ('uniprot', 'Region') ('uniprot', 'Motif') ('uniprot', 'Binding site')  \
0                 False                False                       False   
1                 False                False                       False   
2                 False                False                       False   
3                 False                False                       False   
4                 False                False                       False   

  ('uniprot', 'Glycosylation') HB_intra  ... NB_intra  \
0                        False     True  ...     True   
1                        False     True  ...     True   
2                        False     True  ...     True   
3                        False     True  ...     True   
4                        False     True  ...     True   

  binary_disorder_prediction am_class_pathogenic am_class_benign ss3 ss9  \
0                      False                True           False   H   H   
1                      False                True           False   H   H   
2                      False                True           False   H   H   
3                      False                True           False   H   H   
4                      False                True           False   H   H   

  gene_name family                                       phenotype data_source  
0      HCN1    HCN  Developmental and epileptic encephalopathy, 24     clinvar  
1      HCN1    HCN  Developmental and epileptic encephalopathy, 24     clinvar  
2      HCN1    HCN                        Epileptic encephalopathy     clinvar  
3      HCN1    HCN                        Epileptic encephalopathy     clinvar  
4      HCN1    HCN                                    not provided     clinvar  

[5 rows x 22 columns]

## Initialising the ema object

The `EmbeddingHandler` object is the main object in the ema-tool library. It is used to load the model, preprocess the data, and perform the analysis.
It is initialised with the metadata of the analysed samples.

Embedding spaces are added to the object using the `add_emb_space` method. Multiple embedding spaces can be added to the object. Each embedding space is identified by a unique name. Embedding spaces can also be removed from the object using the `remove_emb_space()` method.

In [10]:
# initialize embedding handler
emb_handler = EmbeddingHandler(metadata)

# add embeddings to the handler
emb_handler.add_emb_space(embeddings=emb_esm1b, emb_space_name='esm1b')
emb_handler.add_emb_space(embeddings=emb_esm2, emb_space_name='esm2')

881 samples loaded.
Categories in meta data: ["('uniprot', 'Topological domain')", "('uniprot', 'Transmembrane')", "('uniprot', 'Intramembrane')", "('uniprot', 'Region')", "('uniprot', 'Motif')", "('uniprot', 'Binding site')", "('uniprot', 'Glycosylation')", 'HB_intra', 'SB_intra', 'DS_intra', 'NB_intra', 'binary_disorder_prediction', 'am_class_pathogenic', 'am_class_benign', 'ss3', 'ss9', 'gene_name', 'family', 'phenotype', 'data_source']
Numerical columns in meta data: ['disorder_propensity']


/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


2 clusters calculated for esm1b.
Embedding space esm1b added.
Embeddings have length 1280.


/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


2 clusters calculated for esm2.
Embedding space esm2 added.
Embeddings have length 1280.


## Accessing the metadata

The sample names can be accessed using the `get_sample_names()` method.

In [11]:
emb_handler.get_sample_names()[:5]

['HCN1:272:S:P',
 'HCN1:279:H:Y',
 'HCN1:391:G:D',
 'HCN1:153:M:I',
 'HCN1:305:M:L']

The columns of the metadata can be accessed shown using the `get_groups()` function. The function returns a list of the column names.

In [12]:
emb_handler.get_groups()

["('uniprot', 'Topological domain')",
 "('uniprot', 'Transmembrane')",
 "('uniprot', 'Intramembrane')",
 "('uniprot', 'Region')",
 "('uniprot', 'Motif')",
 "('uniprot', 'Binding site')",
 "('uniprot', 'Glycosylation')",
 'HB_intra',
 'SB_intra',
 'DS_intra',
 'NB_intra',
 'binary_disorder_prediction',
 'am_class_pathogenic',
 'am_class_benign',
 'ss3',
 'ss9',
 'gene_name',
 'family',
 'phenotype',
 'data_source',
 'cluster_esm1b',
 'cluster_esm2']

The `get_unique_values()` function can be used to get the unique values of a specific column in the metadata.

In [14]:
emb_handler.get_unique_values_per_group(
    group="alphafold2_ss9"
)

ValueError: Column alphafold2_ss9 is not categorical in meta data.

The `get_value_count_per_group` function can be used to get the number of samples per unique value in a specific column.

In [ ]:
emb_handler.get_value_count_per_group(
    group="alphafold2_ss9"
)

The `get_samples_per_group_value()` parameter allows to get the rows of the metadata table that correspond to a specific value in a specific column.

In [ ]:
emb_handler.get_samples_per_group_value(
    group="alphafold2_ss9",
    group_value="B"
)

## Analysis Within an Embedding Space

### Distribution of embedding values

Multiple analyses can be performed to understand the scale and distribution of each embedding space.

This histogram shown with the `plot_emb_hist()` displays the distribution of the embedding values stratified by the embedding space. For an embedding of dimension d=1280 all 1280 values of the embedding will be visualised independently.

In [ ]:
emb_handler.plot_emb_hist()

The `plot_emb_box()` function can be used to display the distribution of the embedding values in a boxplot.

Here the `group` parameter can be used to stratify the samples by a specific column in the metadata or by individual samples.

In [ ]:
emb_handler.plot_emb_box(group="sample")

In [ ]:
emb_handler.plot_emb_box(group="data_source")

### Unsupervised clustering

During the initialisation of the `EmbeddingHandler` object, an unsupervised clustering can be performed automatically using the KMeans algorithm.

By default ema computes a number of clusters equal to the mean number of categories in the metadata. This is a good starting point, but you can also specify the number of clusters you want to compute.


In [ ]:
# specify a number of unsupervised clusters

emb_handler.recalculate_clusters(
    n_clusters=5,
    emb_space_name='esm1b')

emb_handler.recalculate_clusters(
    n_clusters=5,
    emb_space_name='esm2'
    )

The agreement between the clustering and the metadata can be visualised using the `plot_feature_cluster_overlap()` function. This function shows a barplot of the number of samples for each metadata value of the specified metadata column (selected with the `feature` parameter). The colour of the bars represents the cluster assignment of the samples. The height of the bars represents the number of samples in the dataset with the corresponding metadata value.

In [ ]:
emb_handler.plot_feature_cluster_overlap(
    emb_space_name = "esm1b",
    feature="alphafold2_ss3"
)

The function can also be used to compare the agreement between the unsupervised clustering of two different embedding spaces. By setting the `feature` parameter to `cluster_<emb-space-name>` the function will compare the agreement between the unsupervised clustering of the two embedding spaces.

In [ ]:
emb_handler.plot_feature_cluster_overlap(
    emb_space_name = "esm1b",
    feature="cluster_esm2"
)

### Visualisation of dimensionality reduction

Each embedding space can be visualised using dimensionality reduction techniques such as PCA, t-SNE and UMAP. The colour of the points can be stratified by a specific column in the metadata. If the `color` parameter is not set, the points will be coloured by the unsuperivsed clustering.

#### PCA

In [ ]:
emb_handler.visualise_emb_pca(emb_space_name="esm1b",
                              colour="uniprot_region"
                              )

In [ ]:

emb_handler.visualise_emb_pca(emb_space_name="esm2",
                              colour="uniprot_region")

#### t-SNE

In [ ]:
emb_handler.visualise_emb_tsne(emb_space_name="esm1b")

In [ ]:
emb_handler.visualise_emb_umap(
    emb_space_name="esm1b",
    colour="alphamissense_class_benign"
)

### Pairswise distances between samples within an embedding space

The following distance metrics are available for the comparison of pairwise distances between samples:

*   `cosine`: Cosine distance
*   `adjusted_cosine`: Adjusted cosine distance
*   `cityblock`: Manhattan distance
*   `cityblock_normalised`: Normalised manhattan distance
*   `euclidean`: Euclidean distance
*   `sequclidean`: Standardised euclidean distance
*   `sqeuclidean_normalised`: Standardised normalised euclidean distance
*   `euclidean_normalised`: Normalised euclidean distance
*   `mahalanobis`: Mahalanobis distance

The distances are automatically computed as soon as one of the distance functions is called.

All pairwise distances can be accessed in a heatmap. The x and y axes of the heatmap represent the samples. The colour of the heatmap represents the distance between the samples. The samples on each axis can be sorted based on a column in the metadata by using the `x_order` and `y_order` parameters.

In [ ]:
emb_handler.plot_emb_dis_heatmap(
    emb_space_name="esm1b",
    distance_metric="euclidean",
)

The matrix of pairwise distances can be accessed by using the `get_sample_distance()` function. The function returns a NumPy array of the pairwise distances.

In [ ]:
emb_handler.get_sample_distance(
    emb_space_name="esm1b",
    metric="euclidean"
)

The pairwise distances can be normalised with a normal distribution by setting the `rank` parameter to `normal_dis` for all functions that compute pairwise distances. The normalisation is performed by ranking the distances and transforming them to a normal distribution. To inspect whether the pairwise distances in an embedding space are normally distributed, the `plot_emb_dis_his_with_fitted_functions()` function can be used for a visual inspection.

In [ ]:
emb_handler.plot_emb_dis_his_with_fitted_functions(
    emb_space_name="esm1b",
    distance_metric="euclidean",
    rank="normal_dis"
)

### Correlation of pairwise distances with metadata

The correlation of pairwise distances with metadata can be computed using the `plot_emb_dis_continuous_correlation()` function. The function computes the correlation of the pairwise distances with a specific column in the metadata.

In [ ]:
emb_handler.plot_emb_dis_continuous_correlation(
    emb_space_name="esm1b",
    distance_metric="euclidean",
    feature="flDPnn_disorder_propensity"
)

The distances one space can be statified by a specific column in the metadata. The `group` parameter can be used to stratify the samples by a specific column in the metadata.

In [ ]:
emb_handler.plot_emb_dis_per_group(
    emb_space_name="esm1b",
    distance_metric="euclidean",
    group="alphafold2_ss3"
)

The plot can further be stratified by a specific value in the selected column by using the `group_value` parameter.

In [ ]:
emb_handler.plot_emb_dis_per_group(
    emb_space_name="esm1b",
    distance_metric="euclidean",
    group="alphafold2_ss3",
    group_value="H",
    plot_type="box"
)

## Comparing embedding spaces

The distribution of pairwise distances between samples in two embedding spaces can be compared using the `plot_emb_dis_hist` function. The function displays a histogram of the pairwise distances between samples in two embedding spaces.

In [ ]:
emb_handler.plot_emb_dis_hist(distance_metric = "euclidean")

The pairwise distances between two embedding spaces can be visualised using a scatter plot. The `plot_emb_dis_scatter()` function displays the pairwise distances between samples in two embedding spaces. The x and y axes of the scatter plot represent the pairwise distances between samples in the two embedding spaces. The `colour_group` and `colour_value_1` parameters define the column in the metadata and the value in the column which is used to stratify the samples by colour.

In [ ]:
emb_handler.plot_emb_dis_scatter(
    emb_space_name_1 = "esm1b",
    emb_space_name_2 = "esm2",
    distance_metric = "euclidean",
    colour_group = "alphafold2_ss3",
    colour_value_1 = "H",
)

In [ ]:
emb_handler.plot_emb_dis_scatter(
    emb_space_name_1 = "esm1b",
    emb_space_name_2 = "esm2",
    distance_metric = "euclidean",
    colour_group = "alphafold2_ss3",
    colour_value_1 = "H",
    colour_value_2 = "B",
)

The distances can be statified by a specific column in the metadata. The `group` parameter can be used to stratify the samples by a specific column in the metadata.

In [ ]:
emb_handler.plot_emb_dis_box(
    group = "alphafold2_ss3",
    distance_metric="euclidean",
)